src/server/table_hotspot_policy.cpp

```diff
@@ -5,36 +5,110 @@
+        _points[i].write_hotpartition_counter.init_app_counter("app.pegasus",
+                                                               write_counter_name.c_str(),
+                                                               COUNTER_TYPE_NUMBER,
+                                                               write_counter_desc.c_str());
+    }
+}
+
+/*static*/ void hotspot_calculator::notice_replica(const std::string &app_name,
+                                                   const int partition_index,
+                                                   const bool is_read_request)
+{
+    ddebug("start to notice_replica, %s.%d %s",
+           app_name.c_str(),
+           partition_index,
+           is_read_request ? "read_data" : "write_data");
+    std::vector<rpc_address> meta_servers;
+    replica_helper::load_meta_servers(meta_servers);
+    rpc_address meta_server;
+
+    meta_server.assign_group("meta-servers");
+    for (auto &ms : meta_servers) {
+        meta_server.group_address()->add(ms);
     }
+    auto cluster_name = replication::get_current_cluster_name();
+    auto resolver = partition_resolver::get_resolver(cluster_name, meta_servers, app_name.c_str());
+    ::dsn::apps::hotkey_detect_request req;
+    req.type = is_read_request ? dsn::apps::hotkey_type::READ : dsn::apps::hotkey_type::WRITE;
+    req.operation = dsn::apps::hotkey_collector_operation::START;
+    resolver->call_op(RPC_DETECT_HOTKEY,
+                      req,
+                      nullptr,
+                      [app_name, partition_index, is_read_request](
+                          error_code err, dsn::message_ex *request, dsn::message_ex *resp) {
+                          if (err == ERR_OK) {
+                              ::dsn::apps::hotkey_detect_response response;
+                              ::dsn::unmarshall(resp, response);
+                              if (response.err == ERR_OK) {
+                                  ddebug("detect hotspot rpc sending succeed");
+                              } else if (response.err == ERR_SERVICE_ALREADY_EXIST) {
+                                  ddebug("this hotspot rpc has been sending");
+                              } else if (err == ERR_TIMEOUT) {
+                                  ddebug("this hotspot rpc is time_out");
+                              }
+                          }
+                      },
+                      std::chrono::seconds(10),
+                      partition_index,
+                      0);
 }
 
-void hotspot_calculator::start_alg() { _policy->analysis(_app_data, _points); }
+void hotspot_calculator::start_alg()
+{
+    ddebug("start to detect hotspot partition");
+    _policy->analysis(_app_data, _points);
+    if (_hotkey_auto_detect) {
+        for (int i = 0; i < _points.size(); i++) {
+            if (_points[i].read_hotpartition_counter->get_value() >= kHotPartitionT &&
```

Comment:
So you are not checking the topN partitions, but all partitions who exceeds threshold?